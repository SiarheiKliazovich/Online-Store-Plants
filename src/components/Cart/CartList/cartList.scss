@import '../../../sass/variables';
@import '../../../sass/mixins';

.cart__content {
  width: calc(70% - 20px);
	margin-right: 20px;
}

.cart__pag {
	display: flex;
	margin-bottom: 10px;
}

.pag__label {
	font-size: $font-size-lg;
	margin-left: 10px;
}

.pag__select {
	margin: 0 10px;
}

.cart__header {
	display: flex;
	font-size: $font-size-llg;
	border-bottom: 1px solid $color-dark-grey;
	padding-bottom: 10px;
}

.header__name {
	width: 340px;
	margin: 0 60px;
}

.header__price {
	width: 60px;
	margin: 0 10px;
}

.header__quantity {
	width: 100px;
	margin: 0 10px;
}

.header__total {
	width: 100px;
	margin: 0 10px;
}

.cart__items {
	display: flex;
	flex-direction: column;
}

.cart__msg {
	font-size: $font-size-xxlg;
	padding-top: 80px;
	margin: 0 auto;
}

.cart__summary {
	width: 30%;
	background-color: $color-grey;
	border-radius: 10px;
	height: 100%;
	display: flex;
	flex-direction: column;
	align-items: center;
}

.summary__title {
	font-size: $font-size-xxlg;
	padding-top: 40px;
	margin-bottom: 20px;
}

.summary__items {
	font-size: $font-size-xlg;
	margin-bottom: 20px;
}

.summary__amount {
	font-size: $font-size-xlg;
	margin-bottom: 20px;
}

.promo__input {
	border-radius: 5px;
}

.summary__promo {
	display: flex;
	flex-direction: column;
	align-items: center;
}

.promo__input {
	margin-bottom: 5px;
}

.promo__codes {
	margin-bottom: 5px;
}

.summary__promo {
	margin-bottom: 20px;
}

.cart__button {
	margin-bottom: 20px;
	width: 100px;
	height: 40px;
	font-size: $font-size-xlg;
}