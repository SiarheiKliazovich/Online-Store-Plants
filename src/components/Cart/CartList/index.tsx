import { FunctionComponent, useState } from "react";

import products from "../../../data/products";
import CartListItem from "../CartListItem";
import { getUpdateStotage } from "../../../helpers/getUpdateStorage";

import "./cartList.scss";

const CartList = () => {
  let parseCart = getUpdateStotage();
  const [cartData, setCartData] = useState(parseCart);

  const [totalItems, setTotalItems] = useState(0);
  const [totalAmount, setTotalAmount] = useState(0);

  const [selectValue, setSelectValue] = useState(3);
  const [code, setCode] = useState("");

  const updateCart = (id: number, counter: number): void => {
    parseCart = getUpdateStotage();
    const newArrayItems = parseCart.map((item) => {
      if (item.id === id) {
        return { ...item, count: counter };
      }
      return item;
    });
    localStorage.setItem("cart", JSON.stringify(newArrayItems));
  };

  const deleteItem = (id: number): void => {
    parseCart = getUpdateStotage();
    const newArrayItems = parseCart.filter((item) => item.id !== id);
    localStorage.setItem("cart", JSON.stringify(newArrayItems));
    setCartData(newArrayItems);
  };

  const items = parseCart.map((item, i) => {
    const id = item.id;
    return (
      <CartListItem
        key={id}
        {...products[id - 1]}
        i={i}
        deleteItem={deleteItem}
        updateCart={updateCart}
      />
    );
  });

  const selectHandler = (e: React.ChangeEvent<HTMLSelectElement>): void => {
    setSelectValue(+e.target.value);
  };

  const promoCodes = [
    { code: "rs", name: "Rolling Scopes School", amount: 10 },
    { code: "epm", name: "EPAM Systems", amount: 10 },
  ];

  let showAvailableCode = false;
  let availableCode: number;
  const checkPromoCode = (value: string) => {
    promoCodes.map((item, i) => {
      if (item.code === value.toLowerCase()) {
        console.log(1);
        showAvailableCode = true;
        availableCode = i;
      }
    });
  };

  const codeHandler = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setCode(e.target.value);
    checkPromoCode(e.target.value);
  };

  return (
    <>
      <div className="cart__content">
        <div className="cart__pag">
          <div className="pag___items">
            <label className="pag__label">
              Items:
              <select
                className="pag__select"
                value={selectValue}
                onChange={selectHandler}
              >
                <option value="1">1</option>
                <option value="2">2</option>
                <option value="3">3</option>
              </select>
            </label>
          </div>
        </div>
        <div className="cart__header">
          <div className="header__name">Product</div>
          <div className="header__price">Price</div>
          <div className="header__quantity">Quantity</div>
          <div className="header__total">Total</div>
        </div>
        <div className="cart__items">{items}</div>
      </div>
      <div className="cart__summary">
        <div className="summary__title">Summary</div>
        <div className="summary__items">Products: {totalItems}</div>
        <div className="summary__amount">Total: {totalAmount} $</div>
        <div className="summary__promo">
          <div className="promo__input">
            <input
              type="text"
              id="code"
              value={code}
              onChange={(e) => codeHandler(e)}
            />
            {showAvailableCode ? <div>availableCode</div> : ""}
          </div>
          <div className="promo__codes">Promo for test: 'RS', 'EPM'</div>
          <button className="promo__button">Check</button>
        </div>
      </div>
    </>
  );
};

export default CartList;
