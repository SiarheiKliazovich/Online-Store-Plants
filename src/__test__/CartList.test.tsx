import { render } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import CartList from "../components/Cart/CartList";

it("renders cart list", () => {
  const shoppingCart = [{ id: 1, count: 1 }];
  const updateCart = jest.fn();
  const deleteFromCart = jest.fn();
  const sumPrices = 3000;
  const sumCount = 10;
  const setShowModal = jest.fn();
  const cartList = render(
    <Router>
      <CartList
        shoppingCart={shoppingCart}
        updateCart={updateCart}
        deleteFromCart={deleteFromCart}
        sumPrices={sumPrices}
        sumCount={sumCount}
        setShowModal={setShowModal}
      />
    </Router>
  );
  expect(cartList).toMatchSnapshot();
});
