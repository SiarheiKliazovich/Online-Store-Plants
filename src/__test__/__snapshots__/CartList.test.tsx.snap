// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cart list 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cart__content"
      >
        <div
          class="pagination__wrapper"
        >
          <div
            class="cart__title"
          >
            Products In Cart
          </div>
          <div
            class="cart__pag"
          >
            Limit:
            <input
              class="pag__input"
              max="1"
              min="1"
              type="number"
              value="3"
            />
          </div>
          <div
            class="page__number"
          >
            Page:
            <button
              class="next__page"
            >
              &lt;
            </button>
            <span
              class="numder-page"
            >
              1
            </span>
            <button
              class="prev__page"
            >
              &gt;
            </button>
          </div>
        </div>
        <div
          class="cart__header"
        >
          <div
            class="header__name"
          >
            Product
          </div>
          <div
            class="header__price"
          >
            Price
          </div>
          <div
            class="header__quantity"
          >
            Quantity
          </div>
          <div
            class="header__total"
          >
            Total
          </div>
        </div>
        <div
          class="cart__items"
        >
          <li
            class="cart__item"
          >
            <div
              class="cart__item-number"
            >
              1
            </div>
            <a
              href="/product/1"
            >
              <img
                alt="Biotech Garden Jade Plant"
                class="cart__item-img"
                src="https://envatopixel.com/wordpress/plantlish/wp-content/uploads/2021/07/2.webp"
              />
            </a>
            <div
              class="cart__item-desc"
            >
              <a
                href="/product/1"
              >
                <div
                  class="cart__item-text"
                >
                  <div
                    class="cart__item-name"
                  >
                    Biotech Garden Jade Plant
                  </div>
                  <div
                    class="cart__item-category"
                  >
                    Aeonium
                  </div>
                  <div
                    class="cart__item-stock"
                  >
                    In stock: 
                    100
                  </div>
                </div>
              </a>
              <div
                class="cart__price"
              >
                22
                 $
              </div>
              <div
                class="cart__counter"
              >
                <button
                  class="mini-button"
                >
                  -
                </button>
                <div
                  class="counter"
                >
                  1
                </div>
                <button
                  class="mini-button"
                >
                  +
                </button>
              </div>
              <div
                class="cart__total"
              >
                22
                 $
              </div>
            </div>
          </li>
        </div>
      </div>
      <div
        class="cart__summary"
      >
        <div
          class="summary__title"
        >
          Summary
        </div>
        <div
          class="summary__items"
        >
          Products: 
          10
        </div>
        <div
          class="summary__amount"
        >
          Total: 
          3000
          $
        </div>
        <div
          class="summary__promo"
        >
          <div
            class="promo__input"
          >
            <input
              id="code"
              type="search"
              value=""
            />
          </div>
          <div
            class="promo__codes"
          >
            Promo for test: 'RS', 'EPM'
          </div>
        </div>
        <button
          class="cart__button"
        >
          Buy
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cart__content"
    >
      <div
        class="pagination__wrapper"
      >
        <div
          class="cart__title"
        >
          Products In Cart
        </div>
        <div
          class="cart__pag"
        >
          Limit:
          <input
            class="pag__input"
            max="1"
            min="1"
            type="number"
            value="3"
          />
        </div>
        <div
          class="page__number"
        >
          Page:
          <button
            class="next__page"
          >
            &lt;
          </button>
          <span
            class="numder-page"
          >
            1
          </span>
          <button
            class="prev__page"
          >
            &gt;
          </button>
        </div>
      </div>
      <div
        class="cart__header"
      >
        <div
          class="header__name"
        >
          Product
        </div>
        <div
          class="header__price"
        >
          Price
        </div>
        <div
          class="header__quantity"
        >
          Quantity
        </div>
        <div
          class="header__total"
        >
          Total
        </div>
      </div>
      <div
        class="cart__items"
      >
        <li
          class="cart__item"
        >
          <div
            class="cart__item-number"
          >
            1
          </div>
          <a
            href="/product/1"
          >
            <img
              alt="Biotech Garden Jade Plant"
              class="cart__item-img"
              src="https://envatopixel.com/wordpress/plantlish/wp-content/uploads/2021/07/2.webp"
            />
          </a>
          <div
            class="cart__item-desc"
          >
            <a
              href="/product/1"
            >
              <div
                class="cart__item-text"
              >
                <div
                  class="cart__item-name"
                >
                  Biotech Garden Jade Plant
                </div>
                <div
                  class="cart__item-category"
                >
                  Aeonium
                </div>
                <div
                  class="cart__item-stock"
                >
                  In stock: 
                  100
                </div>
              </div>
            </a>
            <div
              class="cart__price"
            >
              22
               $
            </div>
            <div
              class="cart__counter"
            >
              <button
                class="mini-button"
              >
                -
              </button>
              <div
                class="counter"
              >
                1
              </div>
              <button
                class="mini-button"
              >
                +
              </button>
            </div>
            <div
              class="cart__total"
            >
              22
               $
            </div>
          </div>
        </li>
      </div>
    </div>
    <div
      class="cart__summary"
    >
      <div
        class="summary__title"
      >
        Summary
      </div>
      <div
        class="summary__items"
      >
        Products: 
        10
      </div>
      <div
        class="summary__amount"
      >
        Total: 
        3000
        $
      </div>
      <div
        class="summary__promo"
      >
        <div
          class="promo__input"
        >
          <input
            id="code"
            type="search"
            value=""
          />
        </div>
        <div
          class="promo__codes"
        >
          Promo for test: 'RS', 'EPM'
        </div>
      </div>
      <button
        class="cart__button"
      >
        Buy
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
